manager/drivers/secrets.go

```diff
@@ -0,0 +1,51 @@
+package drivers
+
+import (
+	"fmt"
+	"github.com/docker/docker/pkg/plugingetter"
+	"github.com/docker/swarmkit/api"
+)
+
+const (
+	// SecretsProviderAPI is the endpoint for fetching secrets from plugins
+	SecretsProviderAPI = "/SecretProvider.GetSecret"
+
+	// SecretsProviderCapability is the secrets provider plugin capability identification
+	SecretsProviderCapability = "secretprovider"
+)
+
+// SecretDriver provides secrets from different stores
+type SecretDriver struct {
+	plugin plugingetter.CompatPlugin
+}
+
+// NewSecretDriver creates a new driver that provides third party secrets
+func NewSecretDriver(plugin plugingetter.CompatPlugin) *SecretDriver {
+	return &SecretDriver{plugin: plugin}
+}
+
+// Get gets a secret from the secret provider
+func (d *SecretDriver) Get(spec *api.SecretSpec) ([]byte, error) {
+	secretReq := &SecretsProviderRequest{Name: spec.Annotations.Name}
+	var secretResp SecretsProviderResponse
+	err := d.plugin.Client().Call(SecretsProviderAPI, secretReq, &secretResp)
+	if err != nil {
+		return nil, err
+	}
+	if secretResp.Err != "" {
+		return nil, fmt.Errorf(secretResp.Err)
+	}
+	// Assign the secret
+	return []byte(secretResp.Value), nil
+}
+
+// SecretsProviderRequest is the request specification for retrieving secrets from plugins.
+type SecretsProviderRequest struct {
```

Comment:
from discussion in slack: it might be more useful to pass the full `ServiceSpec`, though that would require converting into JSON